File: .opencode/tool/youtube-transcript.ts
import { tool } from "@opencode-ai/plugin"
import { z } from 'zod'
import { YoutubeTranscript } from 't-youtube-transcript-fetcher'
import { sanitizeTitle } from '../utils/sanitize.js'

const DEBUG = process.env.DEBUG === 'yt-processor'

interface YouTubeVideoInfo {
  video_id: string
  title: string
  transcript: string
  description: string
}

const YouTubeApiResponseSchema = z.object({
  items: z.array(z.object({
    snippet: z.object({
      title: z.string(),
      description: z.string()
    })
  }))
})

type YouTubeApiResponse = z.infer<typeof YouTubeApiResponseSchema>

function validateYouTubeVideoId(videoId: string): void {
  if (!videoId || typeof videoId !== 'string') {
    throw new Error('Video ID must be a non-empty string')
  }
  
  // YouTube video IDs are typically 11 characters and contain alphanumeric chars, hyphens, and underscores
  const validIdPattern = /^[a-zA-Z0-9_-]{11}$/
  if (!validIdPattern.test(videoId)) {
    throw new Error('Invalid YouTube video ID format')
  }
}



async function getYouTubeTranscript(videoId: string): Promise<string> {
  validateYouTubeVideoId(videoId)

  if (DEBUG) console.log(`[youtube-transcript] Fetching transcript for video ${videoId}`)

  try {
    // Use the t-youtube-transcript-fetcher library to get transcript
    // This library provides robust transcript fetching with automatic language detection
    const transcriptSegments = await YoutubeTranscript.fetchTranscript(videoId)

    // Convert the array of transcript segments to a single concatenated string
    const transcriptText = transcriptSegments
      .map(segment => segment.text)
      .join(' ')
      .trim()

    if (!transcriptText) {
      throw new Error('No transcript available for this video. The video may not have captions or they may not be accessible through the public API.')
    }

    return transcriptText
  } catch (error) {
    // Handle library-specific errors and convert to our expected error format
    if (error instanceof Error) {
      // Re-throw with our consistent error message for known library errors
      if (error.message.includes('No transcript') ||
          error.message.includes('Transcript disabled') ||
          error.message.includes('Video unavailable') ||
          error.message.includes('Language not found')) {
        throw new Error('No transcript available for this video. The video may not have captions or they may not be accessible through the public API.')
      }
      // For other errors, include the original message
      throw new Error(`Failed to fetch transcript: ${error.message}`)
    }

    // Fallback for unknown error types
    throw new Error('No transcript available for this video. The video may not have captions or they may not be accessible through the public API.')
  }
}

async function getYouTubeMetadata(videoId: string): Promise<{ title: string; description: string }> {
  validateYouTubeVideoId(videoId)
  
  const apiKey = process.env.YOUTUBE_API_KEY
  if (!apiKey) {
    throw new Error('YOUTUBE_API_KEY environment variable not set')
  }
  
  const url = `https://www.googleapis.com/youtube/v3/videos?part=snippet&id=${videoId}&key=${apiKey}`
  
  try {
    const response = await fetch(url)
    if (!response.ok) {
      throw new Error(`Failed to fetch metadata: ${response.statusText}`)
    }
    
    const data = YouTubeApiResponseSchema.parse(await response.json())
    
    if (!data.items || data.items.length === 0) {
      throw new Error(`Video not found: ${videoId}`)
    }
    
    const video = data.items[0].snippet
    const title = sanitizeTitle(video.title)
    
    return {
      title,
      description: video.description
    }
  } catch (error) {
    throw new Error(`Failed to get metadata: ${error instanceof Error ? error.message : 'Unknown error'}`)
  }
}



export default tool({
  description: "Get YouTube video transcript and metadata",
  args: {
    video_id: tool.schema.string().describe("YouTube video ID to fetch transcript and metadata for")
  },
  async execute(args, context): Promise<string> {
    // Validate required environment variable
    const apiKey = process.env.YOUTUBE_API_KEY
    if (!apiKey) {
      throw new Error('YOUTUBE_API_KEY environment variable not set. Please obtain a YouTube Data API v3 key from Google Cloud Console and set the environment variable.')
    }

    try {
      const { title, description } = await getYouTubeMetadata(args.video_id)
      const transcript = await getYouTubeTranscript(args.video_id)

      const result: YouTubeVideoInfo = {
        video_id: args.video_id,
        title,
        transcript,
        description
      }

      return JSON.stringify(result)
    } catch (error) {
      throw new Error(`YouTube tool failed: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  },
}) as any

File: .opencode/tool/youtube_transcript.py
"""YouTube video transcript and metadata tool for OpenCode.

This tool fetches YouTube video metadata and transcripts using the YouTube Data API v3
and youtube-transcript-api library. It provides the same interface as the original
TypeScript implementation but written in Python.
"""

import unicodedata
import json
import os
import re
import sys
from typing import Any

import requests
from dotenv import load_dotenv
from youtube_transcript_api import YouTubeTranscriptApi
from youtube_transcript_api._errors import (
    NoTranscriptFound,
    TranscriptsDisabled,
    VideoUnavailable,
)

# Load environment variables from .env file if it exists
load_dotenv()

DEBUG = os.getenv("DEBUG") == "yt-processor"


def sanitize_title(title: str) -> str:
    """Sanitize a video title for use in filenames and display.

    This function performs the exact same sanitization steps as the original
    TypeScript implementation to ensure consistency.

    Args:
        title: Raw video title from YouTube

    Returns:
        Sanitized title safe for filesystem use

    Examples:
        >>> sanitize_title("Never Gonna Give You Up [Official Video]")
        'Never Gonna Give You Up Official Video'

        >>> sanitize_title("Video: With Special <Characters>")
        'Video- With Special Characters'
    """
    # Normalize Unicode characters (NFC)
    title = unicodedata.normalize("NFC", title)

    # Remove leading brackets [
    title = title.replace("[", "")

    # Remove trailing brackets ]
    title = title.replace("]", "")

    # Replace colons with hyphens
    title = title.replace(":", "-")

    # Remove invalid filename chars and control chars
    # Pattern matches: < > : " / \ | ? * and control characters
    title = re.sub(r'[<>:"/\\|?*\x00-\x1f\x7f-\x9f]', "", title)

    # Remove emojis (basic set)
    title = re.sub(r"[\U0001F600-\U0001F64F]", "", title)

    # Remove symbols & pictographs
    title = re.sub(r"[\U0001F300-\U0001F5FF]", "", title)

    # Remove transport & map symbols
    title = re.sub(r"[\U0001F680-\U0001F6FF]", "", title)

    # Remove flags
    title = re.sub(r"[\U0001F1E0-\U0001F1FF]", "", title)

    # Remove misc symbols
    title = re.sub(r"[\u2600-\u26FF]", "", title)

    # Remove general punctuation
    title = re.sub(r"[\u2000-\u206F]", "", title)

    # Normalize whitespace (collapse to single spaces)
    title = re.sub(r"\s+", " ", title)

    # Trim whitespace
    title = title.strip()

    # Limit length to 100 characters for filesystem compatibility
    title = title[:100]

    return title


def is_safe_filename(title: str) -> bool:
    """Check if a title is safe for use as a filename.

    Args:
        title: Title to check

    Returns:
        True if title is safe, False otherwise
    """
    # Check for invalid filename characters
    if re.search(r'[<>:"/\\|?*\x00-\x1f\x7f-\x9f]', title):
        return False

    # Check for reserved names (Windows)
    reserved_names = {
        "CON",
        "PRN",
        "AUX",
        "NUL",
        "COM1",
        "COM2",
        "COM3",
        "COM4",
        "COM5",
        "COM6",
        "COM7",
        "COM8",
        "COM9",
        "LPT1",
        "LPT2",
        "LPT3",
        "LPT4",
        "LPT5",
        "LPT6",
        "LPT7",
        "LPT8",
        "LPT9",
    }

    if title.upper() in reserved_names:
        return False

    return True


def create_safe_filename(title: str, extension: str = ".md") -> str:
    """Create a safe filename from a title.

    Args:
        title: Video title to convert
        extension: File extension to append (default: .md)

    Returns:
        Safe filename with extension
    """
    sanitized = sanitize_title(title)

    # Ensure we have a valid filename (empty or just whitespace/hyphens)
    if not sanitized or sanitized.strip("-_ ") == "":
        sanitized = "untitled"

    # Add extension
    if not extension.startswith("."):
        extension = "." + extension

    return sanitized + extension


def validate_youtube_video_id(video_id: str) -> None:
    """Validate YouTube video ID format.

    Args:
        video_id: YouTube video ID to validate

    Raises:
        ValueError: If video ID is invalid
    """
    if not video_id or not isinstance(video_id, str):
        raise ValueError("Video ID must be a non-empty string")

    # YouTube video IDs are typically 11 characters and contain alphanumeric chars, hyphens, and underscores
    valid_id_pattern = re.compile(r"^[a-zA-Z0-9_-]{11}$")
    if not valid_id_pattern.match(video_id):
        raise ValueError("Invalid YouTube video ID format")


def get_youtube_transcript(video_id: str) -> str:
    """Fetch YouTube video transcript using youtube-transcript-api.

    Args:
        video_id: YouTube video ID

    Returns:
        Concatenated transcript text

    Raises:
        Exception: If transcript is not available or fetching fails
    """
    validate_youtube_video_id(video_id)

    if DEBUG:
        print(f"[youtube-transcript] Fetching transcript for video {video_id}")

    try:
        # Try to get transcript with language fallbacks
        # First try English variants, then fall back to any available language
        languages = ["en", "en-US", "en-GB", "en-AU"]

        try:
            # Try English variants first
            transcript_segments = YouTubeTranscriptApi().fetch(
                video_id, languages=languages
            )
        except (NoTranscriptFound, TranscriptsDisabled):
            # Fall back to any available transcript
            transcript_segments = YouTubeTranscriptApi().fetch(video_id)

        # Convert the array of transcript segments to a single concatenated string
        transcript_text = " ".join(
            segment.text for segment in transcript_segments
        ).strip()

        if not transcript_text:
            raise Exception(
                "No transcript available for this video. The video may not have captions or they may not be accessible through the public API."
            )

        return transcript_text

    except (NoTranscriptFound, TranscriptsDisabled, VideoUnavailable):
        # Handle library-specific errors and convert to our expected error format
        raise Exception(
            "No transcript available for this video. The video may not have captions or they may not be accessible through the public API."
        )
    except Exception as e:
        # For other errors, include the original message
        raise Exception(f"Failed to fetch transcript: {str(e)}")


def get_youtube_metadata(video_id: str) -> dict[str, str]:
    """Fetch YouTube video metadata using YouTube Data API v3.

    Args:
        video_id: YouTube video ID

    Returns:
        Dictionary containing title and description

    Raises:
        Exception: If metadata fetching fails
    """
    validate_youtube_video_id(video_id)

    api_key = os.getenv("YOUTUBE_API_KEY")
    if not api_key:
        raise Exception(
            "YOUTUBE_API_KEY environment variable not set. Please obtain a YouTube Data API v3 key from Google Cloud Console and set the environment variable."
        )

    url = f"https://www.googleapis.com/youtube/v3/videos?part=snippet&id={video_id}&key={api_key}"

    try:
        response = requests.get(url)
        response.raise_for_status()

        data = response.json()

        if not data.get("items") or len(data["items"]) == 0:
            raise Exception(f"Video not found: {video_id}")

        video = data["items"][0]["snippet"]
        title = sanitize_title(video["title"])

        return {"title": title, "description": video["description"]}

    except requests.RequestException as e:
        raise Exception(f"Failed to get metadata: {str(e)}")
    except Exception as e:
        raise Exception(f"Failed to get metadata: {str(e)}")


def execute(args: dict[str, Any]) -> str:
    """Main tool entry point for OpenCode.

    Args:
        args: Dictionary containing tool arguments
            - video_id: YouTube video ID to fetch transcript and metadata for

    Returns:
        JSON string containing video information

    Raises:
        Exception: If tool execution fails
    """
    # Validate required environment variable
    api_key = os.getenv("YOUTUBE_API_KEY")
    if not api_key:
        raise Exception(
            "YOUTUBE_API_KEY environment variable not set. Please obtain a YouTube Data API v3 key from Google Cloud Console and set the environment variable."
        )

    try:
        video_id = args.get("video_id")
        if not video_id:
            raise Exception("video_id argument is required")

        metadata = get_youtube_metadata(video_id)
        title = metadata["title"]
        description = metadata["description"]
        transcript = get_youtube_transcript(video_id)

        result = {
            "video_id": video_id,
            "title": title,
            "transcript": transcript,
            "description": description,
        }

        return json.dumps(result)

    except Exception as e:
        raise Exception(f"YouTube tool failed: {str(e)}")


# For direct testing
if __name__ == "__main__":
    import sys

    if len(sys.argv) != 2:
        print("Usage: python youtube_transcript.py <video_id>")
        sys.exit(1)

    try:
        result = execute({"video_id": sys.argv[1]})
        print(result)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
